File: 146. LRU Cache.py
class LRUNode:
    def __init__(self, key, value, prev, next):
        self.key = key
        self.value = value
        self.prev = prev
        self.next = next
        
class LRUCache:

    def __init__(self, capacity: int):
        self.head = None
        self.tail = None
        self.capacity = capacity
        self.curr = 0
        self.dic = {}

    def get(self, key: int) -> int:
        if key in self.dic:
            if self.dic[key] is self.head:
                return self.dic[key].value
            elif self.dic[key] is self.tail:
                self.tail = self.tail.prev
                self.tail.next = None
                self.dic[key].next = self.head
                self.dic[key].prev = None
                self.head.prev = self.dic[key]
                self.head = self.dic[key]
            else:
                tn, tp = self.dic[key].next, self.dic[key].prev
                self.dic[key].prev = None
                self.dic[key].next = self.head
                self.head.prev = self.dic[key]
                self.head = self.dic[key]
                tn.prev = tp
                tp.next = tn
            return self.dic[key].value
        else:
            return -1

    def put(self, key: int, value: int) -> None:
        if key in self.dic:
            if self.head is self.dic[key]:
                self.dic[key].value = value
                return
            elif self.tail is self.dic[key]:
                self.tail = self.tail.prev
                self.tail.next = None
                self.dic[key].next = self.head
                self.dic[key].prev = None
                self.head.prev = self.dic[key]
                self.head = self.dic[key]
                self.dic[key].value = value
                return
            else:
                tn, tp = self.dic[key].next, self.dic[key].prev
                self.dic[key].prev = None
                self.dic[key].next = self.head
                self.head.prev = self.dic[key]
                self.head = self.dic[key]
                self.dic[key].value = value
                tn.prev = tp
                tp.next = tn
                return
        else:
            if self.curr < self.capacity:
                if self.curr == 0:
                    node = LRUNode(key, value, None, None)
                    self.head = node
                    self.tail = node
                    self.dic[key] = node
                    self.curr += 1
                else:
                    node = LRUNode(key, value, None, self.head)
                    self.head.prev = node
                    self.head = node
                    self.dic[key] = node
                    self.curr += 1
            else:
                node = LRUNode(key, value, None, self.head)
                self.head.prev = node
                self.head = node
                self.dic.pop(self.tail.key)
                self.tail = self.tail.prev
                self.tail.next = None
                self.dic[key] = node
            return


# Your LRUCache object will be instantiated and called as such:
# obj = LRUCache(capacity)
# param_1 = obj.get(key)
# obj.put(key,value)


from collections import OrderedDict
class LRUCache(OrderedDict):

    def __init__(self, capacity):
        """
        :type capacity: int
        """
        self.capacity = capacity

    def get(self, key):
        """
        :type key: int
        :rtype: int
        """
        if key not in self:
            return - 1
        
        self.move_to_end(key)
        return self[key]

    def put(self, key, value):
        """
        :type key: int
        :type value: int
        :rtype: void
        """
        if key in self:
            self.move_to_end(key)
        self[key] = value
        if len(self) > self.capacity:
            self.popitem(last = False)

# Your LRUCache object will be instantiated and called as such:
# obj = LRUCache(capacity)
# param_1 = obj.get(key)
# obj.put(key,value)

File: 146. LRU Cache.java
/**
 * HashMap + DLL
 */
class LRUCache {
    class Node {
        public int key;
        public int value;
        public Node next;
        public Node prev;
        public Node(int key, int value) {
            this.key = key;
            this.value = value;
        }
    }
    Node head, tail;
    int size;
    int capacity;
    HashMap<Integer, Node> map;

    public LRUCache(int capacity) {
        map = new HashMap<>();
        head = new Node(0,0);
        tail = new Node(0,0);
        head.next = tail;
        tail.prev = head;
        this.capacity = capacity;
        this.size = 0;
    }
    
    public int get(int key) {
        Node n = map.get(key);
        if (n==null) {
            return -1;
        }
        Node prev=n.prev, next=n.next;
        prev.next=next;
        next.prev=prev;
        Node hn = head.next;
        n.prev = head;
        head.next = n;
        hn.prev = n;
        n.next = hn;
        return n.value;
    }
    
    public void put(int key, int value) {
        Node curr = map.get(key);
        if (curr==null) {
            Node newNode = new Node(key,value);
            map.put(key, newNode);
            if (size==capacity) {
                tail = tail.prev;
                map.remove(tail.key);
                size--;
            }  
            Node hn = head.next;
            newNode.prev = head;
            head.next = newNode;
            hn.prev = newNode;
            newNode.next = hn;
            size++;
        } else {
            Node p=curr.prev, n=curr.next;
            p.next = n;
            n.prev = p;
            curr.value = value;
            Node hn = head.next;
            curr.prev = head;
            head.next = curr;
            hn.prev = curr;
            curr.next = hn;
        }
    }
}

/**
 * Your LRUCache object will be instantiated and called as such:
 * LRUCache obj = new LRUCache(capacity);
 * int param_1 = obj.get(key);
 * obj.put(key,value);
 */